initrd/bin/cbfs-init

```diff
@@ -15,17 +15,14 @@ cbfsfiles=`cbfs -t 50 -l 2>/dev/null | grep "^heads/initrd/"`
 for cbfsname in `echo $cbfsfiles`; do
 	filename=${cbfsname:12}
 	if [ ! -z "$filename" ]; then
-		echo "Loading $filename from CBFS"
 		mkdir -p `dirname $filename` \
 		|| die "$filename: mkdir failed"
-		cbfs -t 50 -r $cbfsname > "$filename" \
+		echo "Extracting CBFS file $cbfsname into $filename"
+		cbfs -t 50 $CBFS_ARG -r $cbfsname > "$filename" \
 		|| die "$filename: cbfs file read failed"
 		if [ "$CONFIG_TPM" = "y" ]; then
-			TMPFILE=/tmp/cbfs.$$
-			echo "$filename" > $TMPFILE
-			cat $filename >> $TMPFILE
-			DEBUG "Extending TPM PCR $CONFIG_PCR with $filename"
-			tpmr extend -ix "$CONFIG_PCR" -if $TMPFILE \
+			echo "TPM: Extending PCR[$CONFIG_PCR with] with $filename"
```

Comment:
This used to also extend with the CBFS file name.  I think we should continue to do that:

`tpmr extend -ix "$CONFIG_PCR" -ic "$filename"`

While it's hard to reason about whether an attack could be possible by moving the entire content of one file to another in some way that doesn't invalidate the PCRs, it's easy to just extend with the file name and be sure there's no such attack.

Reply:
@JonathonHall-Purism same comment https://github.com/linuxboot/heads/pull/1758#discussion_r1735138974

Reply:
@JonathonHall-Purism We do not to hash the filenames; we want to hash with file content.

Only occurrences in code of TPM PCR exttends `-ic` ops are to extend PCR[4] with hashes of literals:

```
user@heads-tests-deb12-nix:~/heads$ grep -Rn "\-ic" ./initrd/
./initrd/bin/kexec-select-boot:388:			tpmr extend -ix 4 -ic generic ||
./initrd/bin/usb-init:11:	tpmr extend -ix 4 -ic usb
./initrd/bin/tpmr:241:		-ic)
./initrd/bin/kexec-insert-key:69:tpmr extend -ix 4 -ic generic ||
grep: ./initrd/etc/mtab: No such file or directory
./initrd/etc/ash_functions:245:		tpmr extend -ix 4 -ic recovery
```

Reply:
> We do not to hash the filenames; we want to hash with file content.

It used to hash _both_.  Now it _only_ hashes the content.

This appears to open up an attack by pivoting file contents or renaming files.

For example, say there were only two such files:
* heads/initrd/etc/config.user: `echo CONFIG_RESTRICTED_BOOT=y....`
* heads/initrd/.gnupg/pubring.gpg: `some GPG key data`

An attacker could manipulate CBFS to the following _without affecting the measured PCR state_.
* heads/initrd/etc/config.user.ignore_me_haha: `echo CONFIG_RESTRICTED_BOOT=y...`
* heads/initrd/.gnupg/pubring.gpg.whoops_no_more_gpg: `some GPG key data`

Heads would construct the same PCR measurements in this PR because it is only measuring the file content.  But the restricted boot config and GPG keyring will be gone, since they've been renamed in such a way that they'll be ignored.

Reply:
(should spell this out :slightly_smiling_face: ) - So my point is you do need to hash both the filename and the content.

It used to do this by putting both the filename and content in a temp file, and hashing that.

A better strategy would be to extend once with the filename, then extend a second time with the file content.  (That way the filename and content are clearly separated, you couldn't move the last few letters of the filename to the content, or something like that.)

So include `tpmr extend -ix "$CONFIG_PCR" -ic "$filename"` before hashing the file content.